src/System.Device.Gpio/System/Device/Gpio/GpioController.cs

```diff
@@ -26,7 +26,10 @@ public class GpioController : IDisposable
         private const string HummingBoardProduct = "HummingBoard-Edge";
         private const string HummingBoardHardware = @"Freescale i.MX6 Quad/DualLite (Device Tree)";
 
-        private readonly HashSet<int> _openPins;
+        /// <summary>
+        /// If a pin element exists, that pin is open. Uses current controller's numbering scheme
+        /// </summary>
+        private readonly Dictionary<int, PinValue?> _openPins;
```

Comment:
NIT: I know this wasn't introduced by your change, but since you are changing this, should we make this a ConcurrentDictionary instead? That way it would be safer across threads.

Reply:
I considered this, actually. However, the way it's used right now is not thread safe, even if we use ConcurrentDictionary, because there's always a sequence such as 
```
            if (IsPinOpen(pinNumber))
            {
                throw new InvalidOperationException($"Pin {pinNumber} is already open.");
            }

            OpenPinCore(pinNumber);
            _openPins.Add(pinNumber);
```

Reply:
ConcurrentDictionary won't fail when you call Add more than once across threads. I understand most of this class isn't thread safe, but the idea is to move towards thread-safety not away from it.

Reply:
It's not about ConcurrentDictionary failing, but about calling OpenPinCore twice with the same pin in that scenario. If we want thread safety here, we should use a lock and do it right. Should not hurt much as long as we don't need the lock on `Read` or `Write` operations.